scala-api/src/test/scala/com/cloudera/livy/scalaapi/ScalaClientTest.scala

```diff
@@ -0,0 +1,231 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.cloudera.livy.scalaapi
+
+import java.io._
+import java.net.URI
+import java.nio.charset.StandardCharsets._
+import java.util._
+import java.util.jar.JarOutputStream
+import java.util.zip.ZipEntry
+
+import org.apache.spark.SparkFiles
+import org.apache.spark.launcher.SparkLauncher
+import org.scalatest.{BeforeAndAfter, FunSuite}
+import org.scalatest.concurrent.ScalaFutures
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+import com.cloudera.livy.rsc.RSCConf.Entry._
+
+class ScalaClientTest extends FunSuite with ScalaFutures with BeforeAndAfter {
+
+  import com.cloudera.livy._
+
+  private var client: LivyScalaClient = _
+
+  after {
+    if (client != null) {
+      client.stop(true)
+      client = null
+    }
+  }
+
+  test("test Job Submission") {
+    configureClient(true)
+    val jobHandle = client.submit(ScalaClientTest.helloJob)
+    val result = Await.result(jobHandle, 10 second)
+    assert(result === "hello")
+  }
+
+  test("test Simple Spark Job") {
+    configureClient(true)
+    val sFuture = client.submit(ScalaClientTest.simpleSparkJob)
+    val result = Await.result(sFuture, 10 second)
+    assert(result === 5)
+  }
+
+  test("test Job Failure") {
+    configureClient(true)
+    val sFuture = client.submit(ScalaClientTest.throwExceptionJob)
+    val lock = new Object
+    sFuture onComplete {
+      case Success(t) => {
+        lock.synchronized {
+          lock.notify()
+        }
+        fail("Should have thrown an exception")
+      }
+      case Failure(e) => {
+        lock.synchronized {
+          lock.notify()
+        }
+        assert(e.getMessage.contains("CustomTestFailureException"))
+      }
+    }
+    lock.synchronized {
```

Comment:
This is not thread-safe, since the callback can trigger before you acquire the lock and then you'll miss the notify. It's easier to use a semaphore instead.